crates/wast/src/core/expr.rs

```diff
@@ -1228,42 +1228,31 @@ impl<'a> Parse<'a> for BlockType<'a> {
 pub struct TryTable<'a> {
     pub block: Box<BlockType<'a>>,
     pub catches: Vec<TryTableCatch<'a>>,
-    pub catch_all: Option<TryTableCatchAll<'a>>,
 }
 
 impl<'a> Parse<'a> for TryTable<'a> {
     fn parse(parser: Parser<'a>) -> Result<Self> {
         let block = parser.parse()?;
 
         let mut catches = Vec::new();
-        while parser.peek2::<kw::catch>()? || parser.peek2::<kw::catch_ref>()? {
+        while parser.peek2::<kw::catch>()? || parser.peek2::<kw::catch_ref>()? ||
+              parser.peek2::<kw::catch_all>()? || parser.peek2::<kw::catch_all_ref>()? {
             catches.push(parser.parens(|p| {
                 let kind = if parser.peek::<kw::catch_ref>()? {
                     p.parse::<kw::catch_ref>()?;
-                    TryTableCatchKind::Ref
-                } else {
+                    TryTableCatchKind::CatchRef(p.parse()?)
+                } else if parser.peek::<kw::catch>()? {
                     p.parse::<kw::catch>()?;
-                    TryTableCatchKind::Drop
-                };
-                Ok(TryTableCatch {
-                    kind,
-                    tag: p.parse()?,
-                    label: p.parse()?,
-                })
-            })?);
-        }
-
-        let mut catch_all = None;
-        if parser.peek2::<kw::catch_all>()? || parser.peek2::<kw::catch_all_ref>()? {
-            catch_all = Some(parser.parens(|p| {
-                let kind = if parser.peek::<kw::catch_all_ref>()? {
-                    p.parse::<kw::catch_all_ref>()?;
-                    TryTableCatchKind::Ref
-                } else {
+                    TryTableCatchKind::Catch(p.parse()?)
+                } else if parser.peek::<kw::catch_all>()? {
                     p.parse::<kw::catch_all>()?;
-                    TryTableCatchKind::Drop
+                    TryTableCatchKind::CatchAll
+                } else {
+                    p.parse::<kw::catch_all_ref>()?;
```

Comment:
If you'd like, the [`parser.lookahead1()` method](https://docs.rs/wast/latest/wast/parser/struct.Parser.html#method.lookahead1) can be used here as well to provide a nicer error message if none of the cases match. 

(optional, happy to merge either way)

Reply:
Hmm, so I don't think that would improve the error messages here as we only do these peeks here if the earlier peek2() checks succeeded and we entered the parentheses, so these peeks are infallible and no error will be generated here.

Reply:
Oh good point!